core/src/filed/backup.cc

```diff
@@ -1030,6 +1033,384 @@ static inline bool SendPlainData(b_ctx& bctx)
   return retval;
 }
 
+static result<std::size_t> SendData(BareosSocket* sd,
+                                    POOLMEM* data,
+                                    size_t size)
+{
+  sd->message_length = size;
+  sd->msg = data; /* set correct write buffer */
+
+  if (!sd->send()) {
+    PoolMem error;
+    Mmsg(error, "Network send error to SD. ERR=%s", sd->bstrerror());
+    return error;
+  }
+
+  Dmsg1(130, "Send data to SD len=%d\n", sd->message_length);
+  return size;
+}
+
+class data_message {
+  /* some data is prefixed by a OFFSET_FADDR_SIZE-byte number -- called header
+   * here, which basically contains the file position to which to write the
+   * following block of data.
+   * The difference between FADDR and OFFSET is that offset may be any value
+   * (given to the core by a plugin), whereas FADDR is computed by the core
+   * itself and is equal to the number of bytes already read from the file
+   * descriptor. */
+  static inline constexpr std::size_t header_size = OFFSET_FADDR_SIZE;
+  /* our bsocket functions assume that they are allowed to overwrite
+   * the four bytes directly preceding the given buffer
+   * To keep the message alignment to 8, we "allocate" full 8 bytes instead
+   * of the required 4. */
+  static inline constexpr std::size_t bnet_size = 8;
+  static inline constexpr std::size_t data_offset = header_size + bnet_size;
+
+  std::vector<char> buffer{};
+  bool has_header{false};
+
+ public:
+  data_message(std::size_t data_size)
+  {
+    buffer.resize(data_size + data_offset);
+  }
+  data_message() : data_message(0) {}
+  data_message(const data_message&) = delete;
+  data_message& operator=(const data_message&) = delete;
+  data_message(data_message&&) = default;
+  data_message& operator=(data_message&&) = default;
+
+  // creates a message with the same header -- if any
+  data_message derived() const
+  {
+    data_message derived;
+
+    if (has_header) {
+      derived.has_header = true;
+      std::memcpy(derived.header_ptr(), header_ptr(), header_size);
+    }
+
+    return derived;
+  }
+
+  void set_header(std::uint64_t h)
+  {
+    has_header = true;
+    auto* ptr = header_ptr();
+    std::memcpy(ptr, &h, header_size);
+
+    for (std::size_t i = 0; i < header_size / 2; ++i) {
+      std::swap(ptr[i], ptr[header_size - 1 - i]);
+    }
```

Comment:
that would byteswap on a big-endian platform, too.

Reply:
You are right! Rebased onto master and used the new `network_order` stuff.